README.md

# Slimme Meter Dashboard

Smart meter reading dashboard for Dutch meters with a P1 port.
The dashboard is written in Dutch.

Application is written in NodeJS (API) React (Frontend) Python (Fetcher & Parser)

***The Application is packed in docker containers***

#### Suppported DSMR versions

- DSMR v2.2
- Soon more

## Installation

Use [docker](https://www.docker.com/) for compiling & building the Dashboard environment

```bash
cd /dashboard
docker-compose build && docker-compose up
```

## Usage
After running docker visit ```http://localhost:3000```


## License
[MIT](https://choosealicense.com/licenses/mit/)


frontend/CHANGELOG.md
# Change Log

## [2.0.0] 2021-01-21

### IMPORTANT

- We have updated this product from Bootstrap 3 to Bootstrap 4, so in essence, this is a new product
- For this, we have followed the guidelines from [here](https://react-bootstrap.github.io/migrating/) and [here](https://getbootstrap.com/docs/4.0/migration/)
- We also did not add Bootstrap variables as part of our styles (we will do so in one of our next updates, probably in version 3.0.0 when we'll add Bootstrap 5)

### Bug fixing

- Renamed all files to `.js`, instead of `.jsx`
- Refactor all files from class components to functional ones, so now, you can use React Hooks
- https://github.com/creativetimofficial/light-bootstrap-dashboard-react/issues/62
- https://github.com/creativetimofficial/light-bootstrap-dashboard-react/issues/58
- https://github.com/creativetimofficial/light-bootstrap-dashboard-react/issues/52
- https://github.com/creativetimofficial/light-bootstrap-dashboard-react/issues/50

### Major style changes

- Since the update from Bootstrap 3 to Bootstrap 4, all styles have been changed

### Deleted components

- src/components/Card/Card.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/CustomButton/CustomButton.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/CustomCheckbox/CustomCheckbox.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/CustomRadio/CustomRadio.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/FormInputs/FormInputs.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/StatsCard/StatsCard.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/Tasks/Tasks.jsx (we'll use the simple React Bootstrap tags instead)
- src/components/UserCard/UserCard.jsx (we'll use the simple React Bootstrap tags instead)

### Added components

### Deleted dependencies

- react-notification-system (we'll use react-notification-alert from now on)
- @types/googlemaps
- @types/markerclustererplus
- @types/react
- react-google-maps (we'll use simple Google Maps API with Vanilla JS)
- react-toggle (we'll use the simple React-Bootstrap Form.Check as a switch instead)

### Added dependencies

- react-notification-alert@0.0.13 (instead of the react-notification-system)
- reactstrap@8.9.0 (as part of the react-notification-alert)
- jquery@3.5.1 (since Bootstrap requires it)
- @fortawesome/fontawesome-free@5.15.2 (For easier update of the package - instead of a link inside public/index.html)
- gulp-append-prepend@1.0.8 (For Creative Tim licenses)
- gulp@4.0.2 (For Creative Tim licenses)

### Updated dependencies

```
bootstrap           3.3.7   →    4.6.0
node-sass          4.12.0   →   4.14.1
react              16.8.6   →   17.0.1
react-bootstrap    0.32.4   →    1.4.3
react-chartist     0.13.3   →   0.14.3
react-dom          16.8.6   →   17.0.1
react-router        5.0.0   →    5.2.0
react-router-dom    5.0.0   →    5.2.0
react-scripts       3.0.0   →    4.0.1
typescript          3.4.3   →    4.1.3
```

### Warning

**The TypeScript and jQuery dependencies are installed only to stop console warnings on install. They are not actually used in our product. So the product is not based on TypeScript or jQuery!**
_The following warnings will appear when running the installation command, but they do not affect the UI or the functionality of the product (they will be solved in our next update):_

```
npm WARN react-chartist@0.14.3 requires a peer of react@^0.14.9 || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-popper@1.3.7 requires a peer of react@0.14.x || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN create-react-context@0.3.0 requires a peer of react@^0.14.0 || ^15.0.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
```

_If they will persist in our 2.4.\* version, we will drop their usages and replace them with other plugins._
_In development mode, some of the above plugins will throw a warning because they still use React v16 syntax. If the error will persist in our 2.4.\* version, we will drop their usage and replace them with other plugins._

### Upgrade details

- Updated all packages from `package.json` using `npm-upgrade`, link here: https://www.npmjs.com/package/npm-upgrade
- Renamed all `ControlLabel` components to `FormLabel` (this refers to the react-bootstrap components)
- Renamed all `Grid` components to `Container` (this refers to the react-bootstrap components)
- Rename all `HelpBlock` components to `FormText`
- Rename all `MenuItem` components to `Dropdown.Item` and `NavDropdown.Item`
- Rename all Panel components to Card
- Deleted variables folder
- Change the usage of `React Big Calendar` to the new API, so instead of

```
import BigCalendar from "react-big-calendar";
```

- And

```
const localizer = BigCalendar.momentLocalizer(moment);
```

- We will have:

```
import { Calendar as BigCalendar, momentLocalizer } from "react-big-calendar";
```

- And

```
const localizer = momentLocalizer(moment);
```

- Change Panels with new react-bootstrap components
- For Col componets, replaced `size={number1} sizeOffest={number2}` with `size={{span: number1, offset: number2}}` where size can be `xs`, `sm`, `md`, `lg`
- Changed `bsStyle` to `variant`
- Changed `bsSize` to `size`
- Replace `<Navbar.Header>` with `<div className=“navbar-header”>`
- Replace `<Navbar.Form>` with `<div className="navbar-search-form navbar-form navbar-left">`
- Replace `<FormGroup>` with `<Form.Group>`
- Replace `<InputGroup.Addon>` with `<InputGroup.Prepend><InputGroup.Text>` and `<InputGroup.Append><InputGroup.Text>`
- All required images have a default prop at the end now, example: `const us_flag = require("../assets/img/flags/US.png");` was changed to `const us_flag = require("../assets/img/flags/US.png").default;`
- Rename all .jsx files to .js
- Added Row and Col inside StatsCard

frontend/src/views/Dashboard.js
import React from "react";
import ChartistGraph from "react-chartist";
// react-bootstrap components
import {
  Badge,
  Button,
  Card,
  Navbar,
  Nav,
  Table,
  Container,
  Row,
  Col,
  Form,
  OverlayTrigger,
  Tooltip,
} from "react-bootstrap";

function Dashboard() {
  return (
    <>
      <Container fluid>
        <Row>
          <Col lg="3" sm="6">
            <Card className="card-stats">
              <Card.Body>
                <Row>
                  <Col xs="5">
                    <div className="icon-big text-center icon-danger">
                      <i className="fas fa-tachometer-alt text-danger"></i>
                    </div>
                  </Col>
                  <Col xs="7">
                    <div className="numbers">
                      <p className="card-category">Huidige verbruik</p>
                      <Card.Title as="h4">45 KWh</Card.Title>
                    </div>
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-clock mr-1"></i>
                  Laatste update 13:00 20-06-2021
                </div>
              </Card.Footer>
            </Card>
          </Col>
          <Col lg="3" sm="6">
            <Card className="card-stats">
              <Card.Body>
                <Row>
                  <Col xs="5">
                    <div className="icon-big text-center icon-danger">
                      <i className="fas fa-lightbulb text-danger"></i>
                    </div>
                  </Col>
                  <Col xs="7">
                    <div className="numbers">
                      <p className="card-category">Verbruik Dal & Piek</p>
                      <Card.Title as="h4">50 Kwh</Card.Title>
                    </div>
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-clock mr-1"></i>
                  Laatste update 13:00 20-06-2021
                </div>
              </Card.Footer>
            </Card>
          </Col>
          <Col lg="3" sm="6">
            <Card className="card-stats">
              <Card.Body>
                <Row>
                  <Col xs="5">
                    <div className="icon-big text-center icon-success">
                      <i className="fas fa-leaf text-success"></i>
                    </div>
                  </Col>
                  <Col xs="7">
                    <div className="numbers">
                      <p className="card-category">Teruggave Dal & Piek</p>
                      <Card.Title as="h4">10 Kwh</Card.Title>
                    </div>
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-clock mr-1"></i>
                  Laatste update 13:00 20-06-2021
                </div>
              </Card.Footer>
            </Card>
          </Col>
          <Col lg="3" sm="6">
            <Card className="card-stats">
              <Card.Body>
                <Row>
                  <Col xs="5">
                    <div className="icon-big text-center icon-warning">
                      <i className="fas fa-fire text-warning"></i>
                    </div>
                  </Col>
                  <Col xs="7">
                    <div className="numbers">
                      <p className="card-category">Gas verbruik</p>
                      <Card.Title as="h4">10 m3</Card.Title>
                    </div>
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-clock mr-1"></i>
                  Laatste update 13:00 20-06-2021
                </div>
              </Card.Footer>
            </Card>
          </Col>
        </Row>
        <Row>
          <Col md="8">
            <Card>
              <Card.Header>
                <Card.Title as="h4">Afgelopen maand</Card.Title>
                <p className="card-category">Verbruik & Teruggrave</p>
              </Card.Header>
              <Card.Body>
                <div className="ct-chart" id="chartHours">
                  <ChartistGraph
                    data={{
                      labels: [
                        "9:00AM",
                        "12:00AM",
                        "3:00PM",
                        "6:00PM",
                        "9:00PM",
                        "12:00PM",
                        "3:00AM",
                        "6:00AM",
                      ],
                      series: [
                        [287, 385, 490, 492, 554, 586, 698, 695],
                        [67, 152, 143, 240, 287, 335, 435, 437],
                        [23, 113, 67, 108, 190, 239, 307, 308],
                      ],
                    }}
                    type="Line"
                    options={{
                      low: 0,
                      high: 800,
                      showArea: false,
                      height: "245px",
                      axisX: {
                        showGrid: false,
                      },
                      lineSmooth: true,
                      showLine: true,
                      showPoint: true,
                      fullWidth: true,
                      chartPadding: {
                        right: 50,
                      },
                    }}
                    responsiveOptions={[
                      [
                        "screen and (max-width: 640px)",
                        {
                          axisX: {
                            labelInterpolationFnc: function (value) {
                              return value[0];
                            },
                          },
                        },
                      ],
                    ]}
                  />
                </div>
              </Card.Body>
              <Card.Footer>
                <div className="legend">
                  <i className="fas fa-circle text-danger"></i> Elektrisch
                  Verbruik
                  <i className="fas fa-circle text-info"></i> Elektrisch
                  Teruggave
                  <i className="fas fa-circle text-warning"></i> Gas verbruik
                </div>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-history"></i>
                  Laatste update 13:00 21-06-2021
                </div>
              </Card.Footer>
            </Card>
          </Col>

          <Col md="4">
            <Card>
              <Card.Header>
                <Card.Title as="h4">Verglijking</Card.Title>
                <p className="card-category">Verbruik & Teruggave</p>
              </Card.Header>
              <Card.Body>
                <div
                  className="ct-chart ct-perfect-fourth"
                  id="chartPreferences"
                >
                  <ChartistGraph
                    data={{
                      labels: ["80%", "20%"],
                      series: [80, 20],
                    }}
                    type="Pie"
                  />
                </div>
                <div className="legend">
                  <i className="fas fa-circle text-danger"></i>Verbruik
                  <i className="fas fa-circle text-info"></i>Teruggave
                </div>
                <hr></hr>
                <div className="stats">
                  <i className="far fa-clock"></i>
                  Laatste update 13:00 21-06-2021
                </div>
              </Card.Body>
            </Card>
          </Col>
        </Row>
        <Row>
          <Col md="6">
            <Card>
              <Card.Header>
                <Card.Title as="h4">Jaar overzicht</Card.Title>
                <p className="card-category">Jaar overzicht verbruik</p>
              </Card.Header>
              <Card.Body>
                <div className="ct-chart" id="chartActivity">
                  <ChartistGraph
                    data={{
                      labels: [
                        "Jan",
                        "Feb",
                        "Mar",
                        "Apr",
                        "Mai",
                        "Jun",
                        "Jul",
                        "Aug",
                        "Sep",
                        "Oct",
                        "Nov",
                        "Dec",
                      ],
                      series: [
                        [
                          542, 443, 320, 780, 553, 453, 326, 434, 568, 610, 756,
                          895,
                        ],
                        [
                          412, 243, 280, 580, 453, 353, 300, 364, 368, 410, 636,
                          695,
                        ],
                        [
                          512, 243, 42, 580, 453, 353, 300, 364, 368, 410, 636,
                          695,
                        ],
                      ],
                    }}
                    type="Bar"
                    options={{
                      seriesBarDistance: 10,
                      axisX: {
                        showGrid: false,
                      },
                      height: "245px",
                    }}
                    responsiveOptions={[
                      [
                        "screen and (max-width: 640px)",
                        {
                          seriesBarDistance: 5,
                          axisX: {
                            labelInterpolationFnc: function (value) {
                              return value[0];
                            },
                          },
                        },
                      ],
                    ]}
                  />
                </div>
              </Card.Body>
              <Card.Footer>
                <div className="legend">
                  <i className="fas fa-circle text-info"></i> Energie Verbruik
                  <i className="fas fa-circle text-danger"></i> Energie
                  Teruggave
                  <i className="fas fa-circle text-warning"></i> Gas verbruik
                </div>
                <hr></hr>
                <div className="stats">
                  <i className="fas fa-check"></i>
                  Data information certified
                </div>
              </Card.Footer>
            </Card>
          </Col>
        </Row>
      </Container>
    </>
  );
}

export default Dashboard;
